rtCommon/bidsArchive.py

```diff
@@ -0,0 +1,825 @@
+        dimensions2 = header2.get("dim")
+
+        nDimensions1 = dimensions1[0]
+        nDimensions2 = dimensions2[0]
+
+        pixdim1 = header1.get("pixdim")
+        pixdim2 = header2.get("pixdim")
+
+        # Case 1
+        if nDimensions1 == nDimensions2:
+            pixdimEqual = np.array_equal(pixdim1, pixdim2)
+            allButFinalEqual = np.array_equal(dimensions1[:nDimensions1],
+                                              dimensions2[:nDimensions2])
+
+            if not (pixdimEqual and allButFinalEqual):
+                dimensionMatch = False
+        # Case 2
+        else:
+            dimensionsDifferBy1 = abs(nDimensions1 - nDimensions2) == 1
+
+            nSharedDimensions = min(nDimensions1, nDimensions2)
+            # Arrays are 1-indexed as # dimensions is stored in first slot
+            sharedDimensionsMatch = \
+                np.array_equal(dimensions1[1:nSharedDimensions + 1],
+                               dimensions2[1:nSharedDimensions + 1])
+            # Arrays are 1-indexed as value used in one method of voxel-to-world
+            # coordination translation is stored in the first slot (value should
+            # be equal across images)
+            sharedPixdimMatch = np.array_equal(pixdim1[:nSharedDimensions + 1],
+                                               pixdim2[:nSharedDimensions + 1])
+
+            if not (dimensionsDifferBy1 and sharedDimensionsMatch and
+                    sharedPixdimMatch):
+                dimensionMatch = False
+
+        if not dimensionMatch:
+            errorMsg = ("NIfTI headers not append compatible due to mismatch "
+                        "in dimensions and pixdim fields.\n"
+                        f"Dim 1: {dimensions1} | Dim 2: {dimensions2}\n"
+                        f"Pixdim 1: {pixdim1} | Pixdim 2: {pixdim2}\n")
+            return (False, errorMsg)
+
+        # Compare xyzt_units (spatial and temporal dimension units)
+        # Spatial and temporal dimensions should both be the same; however, it's
+        # valid for only one image to have temporal dimensions defined, as long
+        # as spatial is the same for both
+        field = 'xyzt_units'
+        xyztUnits1 = header1[field]
+        xyztUnits2 = header2[field]
+
+        spatialUnits1 = xyztUnits1 % 8
+        spatialUnits2 = xyztUnits2 % 8
+        spatialMatch = np.array_equal(spatialUnits1, spatialUnits2)
+
+        temporalUnits1 = xyztUnits1 - spatialUnits1
+        temporalUnits2 = xyztUnits2 - spatialUnits2
+        temporalMatch = np.array_equal(temporalUnits1, temporalUnits2)
+
+        # Append compatible if both dimensions match
+        if spatialMatch and temporalMatch:
+            pass
+        # Append compatible if spatial dimensions match, temporal dimensions
+        # don't, and at least one temporal dimension is undefined
+        elif (spatialMatch and not temporalMatch and
+              (temporalUnits1 == 0 or temporalUnits2 == 0)):
```

Comment:
GW comment: do we need to check that at least one of them != 0?

Reply:
TODO: Document the magic number 0

Reply:
TODO: Document more on the philosophy of 'if things are defined, and they don't match, then the headers don't match; otherwise, they are let through"